File: 01/01.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func readFile(name string) []int {
	file, err := os.Open(name)
	if err != nil {
		log.Fatal("failed to open")
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	var items []int

	for scanner.Scan() {
		item, _ := strconv.Atoi(scanner.Text())
		items = append(items, item)
	}

	return items
}

func interval(gaps []int, gap int) int {
	greater := 0
	for i := 0; i < len(gaps)-gap; i++ {
		if gaps[i+gap] > gaps[i] {
			greater++
		}
	}
	return greater
}

func main() {
	text := readFile("input")
	fmt.Println(interval(text, 1)) // 1548
	fmt.Println(interval(text, 3)) // 1589
}

File: 01/01.cc
#include <fstream>
#include <iostream>
#include <vector>

template <class T> auto read_file(const std::string &name) {
  auto file = std::ifstream(name);
  auto items = std::vector<T>{};
  auto item = T{};
  while (file >> item)
    items.emplace_back(item);
  return items;
}

template <class T> auto interval(const std::vector<T> &gaps, T gap) {
  auto greater = 0;
  for (auto i = size_t{}, end = gaps.size() - gap; i < end; ++i)
    if (gaps[i + gap] > gaps[i])
      ++greater;
  return greater;
}

int main() {
  auto items = read_file<int>("input");
  std::cout << interval(items, 1) << '\n'  // 1548
            << interval(items, 3) << '\n'; // 1589
}
